kinto/core/resource/__init__.py

```diff
@@ -637,6 +663,8 @@ def apply_changes(self, record, changes):
         content_type = str(self.request.headers.get('Content-Type'))
         if content_type == 'application/merge-patch+json':
             recursive_update_dict(updated, changes, ignores=[None])
+        elif content_type == 'application/json-patch+json':
+            updated = self.patch_changes.updated
```

Comment:
This could be moved on top, before iterating on `changes` (see previous comment about *current flow*).
With a new dedicated helper:

```py
if content_type == 'application/json-patch+json':
    try:
        updated  = apply_json_patch(updated, changes)
    except ValueError as e:
        raise_invalid(...)
```